{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1: Introduction to SpatialData & Interactive Exploration\n",
    "\n",
    "**Tutor:** Tim Treis\n",
    "**Time:** 40 minutes\n",
    "\n",
    "---\n",
    "\n",
    "Welcome to the workshop! In this first notebook, we will introduce the core concepts behind the `SpatialData` framework.\n",
    "\n",
    "**Goals:**\n",
    "1. Understand the different components of a spatial dataset (Images, Labels, Shapes, Points).\n",
    "2. Learn how the `SpatialData` object acts as a unified container for these components.\n",
    "3. Gain hands-on experience exploring complex spatial data interactively with `napari`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is a `SpatialData` object?\n",
    "\n",
    "Modern spatial omics experiments generate diverse types of data. For example, a single experiment might produce:\n",
    "\n",
    "*   A high-resolution histology image (**Image**).\n",
    "*   Segmentation masks defining where the cells are (**Labels**).\n",
    "*   Polygons outlining cell boundaries (**Shapes**).\n",
    "*   The locations of individual RNA molecules (**Points**).\n",
    "*   A table of gene counts per cell (**Table**).\n",
    "\n",
    "The `SpatialData` framework, part of the `scverse` ecosystem, provides a single, standardized object to hold all these different *elements* together in a coordinated way.\n",
    "\n",
    "<img src=\"../resources/elements.png\" alt=\"spatialdata design is modular\" style=\"max-width: 800px;\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading our First Dataset: A Xenium Example\n",
    "\n",
    "To begin, we will load a pre-processed dataset into a `SpatialData` object. For this workshop, we will be working with the `.zarr` format, which is a modern, high-performance storage format ideal for large scientific data.\n",
    "\n",
    "Our first dataset is a `Xenium` experiment, which is a great example because it contains all the different types of spatial elements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext jupyter_black\n",
    "\n",
    "import spatialdata as sd\n",
    "\n",
    "# Define the path to our data directory\n",
    "# Note: This path is relative to the repository's root directory\n",
    "data_path = \"../data/\"\n",
    "\n",
    "print(data_path + \"xenium_lung_cancer_subset.zarr\")\n",
    "\n",
    "# Load the pre-processed Xenium dataset\n",
    "sdata_xenium = sd.read_zarr(data_path + \"xenium_lung_cancer_subset.zarr\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's inspect the `SpatialData` object we just created. Notice how it contains all the different data elements, including **multiscale images**. These are image pyramids that store the same image at different resolutions, which is key for fast visualization of very large images.\n",
    "\n",
    "![image pyramid](../resources/image_pyramid.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata_xenium"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can easily access any part of the object, for example, the table of gene counts, which is stored as an `anndata.AnnData` object that many of you may be familiar with from single-cell analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata_xenium.tables[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Interactive Exploration with `napari`\n",
    "\n",
    "While printing the object summary is useful, the best way to understand spatial data is to *see* it. We use **Napari**, a fast, interactive, multi-dimensional image viewer, to explore our `SpatialData` objects.\n",
    "\n",
    "The `napari-spatialdata` plugin provides the bridge between our data and the viewer. To demonstrate its features, we'll load a `MERFISH` dataset, another high-resolution imaging-based technology."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata_merfish = sd.read_zarr(data_path + \"merfish_subset.zarr\")\n",
    "sdata_merfish"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Your Turn: Launching Napari\n",
    "\n",
    "Now, let's launch the interactive viewer. This is the first major hands-on exercise of the workshop.\n",
    "\n",
    "**Instructions:**\n",
    "1. Uncomment and run the code cell below.\n",
    "2. A new window for Napari should appear. This may take a few seconds.\n",
    "3. Follow along as we tour the interface and explore the data layers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from napari_spatialdata import Interactive\n",
    "\n",
    "# uncomment the line below to launch napari\n",
    "Interactive(sdata_merfish)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"border: 1px solid #4CAF50; border-left-width: 15px; padding: 10px; background-color: #F0FFF0; color: black;\">\n",
    "    <strong>Tip:</strong>\n",
    "    <p>You can also view a .zarr file from your terminal with:<br><code>python -m napari_spatialdata view ../data/merfish_subset.zarr</code>.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tour of the `napari-spatialdata` Interface\n",
    "\n",
    "The Napari window is composed of several key parts. The `napari-spatialdata` plugin adds the panels highlighted in green.\n",
    "\n",
    "![overview of the napari-spatialdata interface](../resources/napari_spatialdata0.jpg)\n",
    "\n",
    "**Key areas to know:**\n",
    "- **Layers list (bottom-left):** This is where you can see all the data layers (images, shapes, points) and toggle their visibility, change their order, or adjust their opacity.\n",
    "- **`spatialdata` viewer (top-right):** This shows you which coordinate systems are available and lets you select elements to view.\n",
    "- **`spatialdata` table annotation (bottom-right):** This powerful panel lets you color your spatial elements (like cells) by any value in the associated `AnnData` table."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is an example of adding elements to the viewer and interacting with them.\n",
    "\n",
    "![adding elements](../resources/napari_spatialdata1.gif)\n",
    "![interaction with napari](../resources/napari_spatialdata2.gif)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The real power comes from coloring your spatial elements by their associated data. Here, we color cells by different annotations stored in the `AnnData` table, including categorical variables (`region`) and continuous variables (gene expression).\n",
    "\n",
    "![showing annotations](../resources/napari_spatialdata3.gif)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The various selectors on the right side of the interface reflect the structure of the `AnnData` object that annotates our spatial elements. The dropdown menus map directly to the `.obs`, `.layers`, and `.obsm` slots of the table.\n",
    "\n",
    "<img src=\"https://raw.githubusercontent.com/scverse/anndata/main/docs/_static/img/anndata_schema.svg\" width=\"500\" height=\"500\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Advanced Tip: Coloring Points by their Own Attributes\n",
    "\n",
    "Most of the time, you will color cells/spots by data in an `AnnData` table. However, it's also possible to color `Points` or `Shapes` by columns in their own internal dataframe. For example, individual transcripts might have a `quality_value` column.\n",
    "\n",
    "The \"Dataframe columns\" list widget at the bottom of the annotation panel allows you to do this.\n",
    "\n",
    "![showing annotations points](../resources/napari_spatialdata4.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This concludes our introduction to the `SpatialData` object and interactive visualization. In the next notebook, we'll learn how to create static, publication-ready plots and compute our first spatial statistics."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
