{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2: Plotting and Basic Spatial Metrics\n",
    "\n",
    "**Instructor:** Tim Treis\n",
    "**Time:** 35 minutes\n",
    "\n",
    "---\n",
    "\n",
    "Now that we understand what a `SpatialData` object is and how to explore it interactively, let's learn how to create static, publication-quality plots and ask our first spatial question: \"Is this gene's expression pattern random, or is it spatially organized?\"\n",
    "\n",
    "**Goals:**\n",
    "1. Use `spatialdata-plot` to create layered, static images.\n",
    "2. Introduce a different technology: spot-based Visium data.\n",
    "3. Use `squidpy` to calculate and visualize a key spatial statistic, Moran's I."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup\n",
    "First, let's import the libraries we'll need for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext jupyter_black\n",
    "\n",
    "import spatialdata as sd\n",
    "import spatialdata_plot as sdp\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "\n",
    "data_path = \"../data/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Static Plotting with `spatialdata-plot`\n",
    "\n",
    "For this section, we will use a dataset from a 10x Genomics Visium experiment. Unlike the high-resolution Xenium data from the previous notebook, Visium is a **spot-based** technology. It captures the whole transcriptome, but at the resolution of spots (~55Âµm) which may contain multiple cells.\n",
    "\n",
    "![Visium technology](../resources/visium_tech.png)\n",
    "\n",
    "Let's load our pre-processed Visium dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata_visium = sd.read_zarr(data_path + \"visium_glioblastoma_subset.zarr\")\n",
    "sdata_visium"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `spatialdata-plot` library adds a `.pl` accessor to our `SpatialData` object. We can chain `render_*` functions to build up a plot layer by layer, similar to `ggplot`.\n",
    "\n",
    "Let's start by rendering the histology image."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata_visium.pl.render_images().pl.show(coordinate_systems=\"downscaled_hires\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's overlay the circular Visium spots (`Shapes`) on top of the image. We can make them semi-transparent to see the tissue underneath."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    sdata_visium.pl.render_images()\n",
    "    .pl.render_shapes(fill_alpha=0.3)\n",
    "    .pl.show(\"downscaled_hires\", figsize=(8, 8))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identifying Spatially Variable Genes\n",
    "\n",
    "Simply seeing the spots isn't very informative. A common first question in spatial analysis is to identify genes that show a non-random spatial pattern. Are some genes expressed only in one corner of the tissue? \n",
    "\n",
    "To answer this, we can use tools from `squidpy` to calculate a spatial autocorrelation score called **Moran's I**.\n",
    "\n",
    "![spatial patterns](../resources/spatial_patterns.png)\n",
    "\n",
    "*   A **high positive** Moran's I (near +1) means the gene is **clustered** spatially.\n",
    "*   A Moran's I **near 0** means the gene expression is **random**.\n",
    "*   A **high negative** Moran's I (near -1) means the gene shows a **dispersed** or checkerboard-like pattern."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To calculate Moran's I, we first need to build a **spatial graph**. This graph connects spots that are neighbors in physical space, which allows `squidpy` to compare their expression values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the AnnData table from our SpatialData object\n",
    "adata_visium = sdata_visium.tables[\"table\"]\n",
    "\n",
    "# Build the spatial neighborhood graph\n",
    "sq.gr.spatial_neighbors(adata_visium)\n",
    "\n",
    "# Calculate Moran's I for all genes\n",
    "sq.gr.spatial_autocorr(\n",
    "    adata_visium,\n",
    "    mode=\"moran\",\n",
    "    n_perms=100,  # Use more permutations for a real analysis\n",
    "    n_jobs=4,  # Use multiple cores to speed it up\n",
    ")\n",
    "\n",
    "# Let's look at the genes with the highest Moran's I score (most clustered)\n",
    "adata_visium.uns[\"moranI\"].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can use `spatialdata-plot` to visualize the expression of one of these top-scoring genes. We expect to see a clear spatial pattern.\n",
    "\n",
    "Let's plot the expression of `CST3`, which has a very high score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 2, figsize=(12, 5))\n",
    "\n",
    "# Plot the H&E image on the left\n",
    "sdata_visium.pl.render_images().pl.show(\n",
    "    \"downscaled_hires\", ax=axs[0], title=\"H&E Image\"\n",
    ")\n",
    "\n",
    "# Plot the gene expression on the right by coloring the spots\n",
    "(\n",
    "    sdata_visium.pl.render_images()\n",
    "    .pl.render_shapes(color=\"CST3\", size=250)  # Color spots by CST3 expression\n",
    "    .pl.show(\n",
    "        \"downscaled_hires\", ax=axs[1], title=\"CST3 Expression (Spatially Variable)\"\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see, the gene expression is highly concentrated in a specific region of the tissue, confirming the high Moran's I score.\n",
    "\n",
    "This concludes the first part of our workshop. We've learned how to load and interact with different spatial technologies and how to perform basic plotting and spatial statistical analysis. In the next part, Anthony will guide us through a complete downstream analysis, including cell type clustering and integrating data from multiple experiments."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
